ch03tests/05Mocks.ipynb

```diff
@@ -4,54 +4,75 @@
    "cell_type": "markdown",
```

Comment:
<p>What's the reason to make <code>params</code> an argument for the function? I see you've added an example with a different set of parameters below. But tin the example we are building a test and the parameters need to be the good ones, the ones we want to check the call is being done so when the function is called that way. In other words, the defaults to the <code>map_at</code> function. Showing two set of <code>params</code> in this example diverts the understanding of what's being tested.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/UCL/rsd-engineeringcourse/pull/221/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='28'/>

Reply:
Thanks @dpshelio, 
I think the example in the notebook is testing the use of the right parameters to do a right remote call: " Instead, we can use mocks to assert that our code does the right thing in terms of the messages it sends: the parameters of the function calls it makes to the remote resource". Just amended the cell to show this:

```
def test_build_default_params(_base, _params):
    with patch.object(requests,'get') as mock_get:
        default_map = map_at(51.0, 0.0)
        mock_get.assert_called_with(_base, params=_params)

base="https://static-maps.yandex.ru/1.x/?"
good_params={
    'z':12,
    'size':'400,400',
    'll':'0.0,51.0',
    'lang':'en_US',
    'l': 'map'
}        
test_build_default_params(base, good_params)
```
and

```
base="https://static-maps.yandex.ru/1.x/?"
bad_params={
    'z':15,
    'size':'400,400',    
    'll':'0.0,51.0',
    'lang':'en_US',
    'l': 'map'
}        
test_build_default_params(base, bad_params)
```
In any case, as I am new to this kind of testing. Examples require to be more intuitive and perhaps add more details for patching as there are few handy tests (e.g. https://kimsereylam.com/python/2021/03/19/how-to-use-patch-in-python-unittest.html). Happy to chat about this bits with @alessandrofelder and yourself next Tuesday :) 

Thanks.

Reply:
I think a test called `test_build_default_params` should only test the default parameters. But I like the idea of showing the versatility of mocking. Could we have an additional bit of text after `test_build_default_params` that says something like "Mocking can also be used to test how we handle errors we receive from the remote resource" and then have a negative test that checks we handle a mocked dodgy internet connection as expected?